src/category/fold.lean

```diff
@@ -0,0 +1,344 @@
+
+universes u v
+
+open ulift category_theory
+
+namespace monoid
+variables {m : Type u → Type u} [monad m]
+variables {α β : Type u}
+/--
+For a list, foldl f x [y₀,y₁] reduces as follows
+  calc  foldl f x [y₀,y₁]
+      = foldl f (f x y₀) [y₁]      : rfl
+  ... = foldl f (f (f x y₀) y₁) [] : rfl
+  ... = f (f x y₀) y₁              : rfl
+
+with f : α → β → α
+     x : α
+     [y₀,y₁] : list β
+
+We can view the above as a composition of functions:
+
+  ... = f (f x y₀) y₁             : rfl
+  ... = flip f y₁ (flip f y₀ x)   : rfl
+  ... = (flip f y₁ ∘ flip f y₀) x : rfl
+
+We can use traverse and const to construct this composition:
+
+  calc   const.run (traverse (λ y, const.mk' (flip f y)) [y₀,y₁]) x
+       = const.run ((::) <$> const.mk' (flip f y₀) <*> traverse (λ y, const.mk' (flip f y)) [y₁]) x
+  ...  = const.run ((::) <$> const.mk' (flip f y₀) <*> ( (::) <$> const.mk' (flip f y₁) <*> traverse (λ y, const.mk' (flip f y)) [] )) x
+  ...  = const.run ((::) <$> const.mk' (flip f y₀) <*> ( (::) <$> const.mk' (flip f y₁) <*> pure [] )) x
+  ...  = const.run ( ((::) <$> const.mk' (flip f y₁) <*> pure []) ∘ ((::) <$> const.mk' (flip f y₀)) ) x
+  ...  = const.run ( const.mk' (flip f y₁) ∘ const.mk' (flip f y₀) ) x
+  ...  = const.run ( flip f y₁ ∘ flip f y₀ ) x
+  ...  = f (f x y₀) y₁
+
+And this is how `const` turns a monoid into an applicative functor and
+how the monoid of endofunctions define `foldl`.
+-/
+@[reducible] def foldl (α : Type u) : Type u := End α
+def foldl.mk (f : α → α) : foldl α := f
+def foldl.of_free_monoid (f : β → α → β) (xs : free_monoid α) : monoid.foldl β :=
+flip (list.foldl f) xs
+
+@[reducible] def foldr (α : Type u) : Type u := (End α)ᵒᵖ
+def foldr.mk (f : α → α) : foldr α := op f
+def foldr.get (x : foldr α) : α → α := unop x
+def foldr.of_free_monoid (f : α → β → β) (xs : free_monoid α) : monoid.foldr β :=
+op $ flip (list.foldr f) xs
+
+@[reducible] def mfoldl (m : Type u → Type u) [monad m] (α : Type u) : Type u := End $ Kleisli.mk m α
+def mfoldl.mk (f : α → m α) : mfoldl m α := f
+def mfoldl.of_free_monoid (f : β → α → m β) (xs : free_monoid α) : monoid.mfoldl m β :=
+flip (list.mfoldl f) xs
+
+@[reducible] def mfoldr (m : Type u → Type u) [monad m] (α : Type u) : Type u := (End $ Kleisli.mk m α)ᵒᵖ
+def mfoldr.mk (f : α → m α) : mfoldr m α := op f
+def mfoldr.get (x : mfoldr m α) : α → m α := unop x
+def mfoldr.of_free_monoid (f : α → β → m β) (xs : free_monoid α) : monoid.mfoldr m β :=
+op $ flip (list.mfoldr f) xs
+
+end monoid
+
+namespace traversable
+open monoid functor
+
+section defs
+variables {α β : Type u} {t : Type u → Type u} [traversable t]
+
+def fold_map {α ω} [has_one ω] [has_mul ω] (f : α → ω) : t α → ω :=
+traverse (const.mk' ∘ f)
+
+def foldl (f : α → β → α) (x : α) (xs : t β) : α :=
+fold_map (foldl.mk ∘ flip f) xs x
+
+def foldr (f : α → β → β) (x : β) (xs : t α) : β :=
+unop (fold_map (foldr.mk ∘ f) xs) x
+
+/--
+`to_list_spec` provides an alternative definition for `to_list`:
```

Comment:
The fact that this comment starts off with `to_list_spec` but is immediately followed by `def to_list` is confusing.

Reply:
What do you think about swapping the two paragraphs?

Reply:
I'm still a bit confused. Which is “the definition based on `foldr`” that you are referring to?

Reply:
Oh! I see! That's outdated documentation. It should say `foldl`

Reply:
How about:

/--
Conceptually, `to_list` collects all the elements of a collection
in a list. This idea is formalized by 

  `lemma to_list_spec (x : t α) : to_list x = fold_map free_monoid.mk x`.

The definition of `to_list` is based on `foldl` and `list.cons` for
speed. It is faster than using `fold_map free_monoid.mk` because, by
using `foldl` and `list.cons`, each insertion is done in constant
time. As a consequence, `to_list` performs in linear.

On the other hand, `fold_map free_monoid.mk` creates a singleton list
around each element and concatenates all the resulting lists. In 
`xs ++ ys`, concatenation takes a time proportional to `length xs`. Since
the order in which concatenation is evaluated is unspecified, nothing
prevents each element of the traversable to be appended at the end 
`xs ++ [x]` which would yield a `O(n²)` run time. -/